### PR TITLE
1342 move recaptcha and lockout

### PR DESCRIPTION
**Pull request by nasief**

### CODE REVIEW

**Comment by nasief** [comment]
[(2) 53TWO - Move login lockout and recaptcha to django-users-plus [Week42]](https://trello.com/c/TiQlAPZC/1342-2-53two-move-login-lockout-and-recaptcha-to-django-users-plus-week42)

File: README.rst [hunk 1]
```diff
@@ -34,6 +34,8 @@ Quick start
         ...
         'accountsplus',
         'django.contrib.sites',
+        'captcha',
+        'axes',
```
Comment (waseem-omar) [inline README.rst:39]: @nasief I think we need to add `timezone_field` to installed apps as well.

File: README.rst [hunk 2]
```diff
@@ -69,3 +71,21 @@ Quick start
...
+
+        RECAPTCHA_PUBLIC_KEY = <Key generated from Google Recaptcha API>
+        RECAPTCHA_PRIVATE_KEY = <Key generated from Google Recaptcha API>
+        RECAPTCHA_USE_SSL = True
+
+        AXES_LOCK_OUT_AT_FAILURE = True
+        AXES_COOLOFF_TIME = <Integer number of hours to block user>
+        AXES_LOGIN_FAILURE_LIMIT = <Count of failing attempts before lockout>
+        AXES_LOCKOUT_URL = 'locked/'
+        AXES_LOCKOUT_TEMPLATE = <Template you want to show on lockout>
```
Comment (waseem-omar) [inline README.rst]: @nasief Please add references to `django-axes` and `django-recaptcha` libraries' documentations.

File: accountsplus/settings.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
+from django.conf import settings
+from django.core.exceptions import ImproperlyConfigured
+from django.apps import apps
+
+
+def get_setting(setting_str, default_value=None):
+    try:
+        return getattr(settings, setting_str, default_value)
```
Comment (waseem-omar) [inline accountsplus/settings.py]: @nasief This will never raise an error because we are passing default_value parameter.

Comment (waseem-omar) [reply accountsplus/settings.py]: @nasief You may want to add another parameter to `get_setting` method to check if this setting is required or not, and pass a default value only if the setting you are trying to get is optional.

File: accountsplus/settings.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+    # Checking if recaptcha is enabled over SSL
+    RECAPTCHA_USE_SSL = bool(is_ssl_enabled_for_captcha())
+    if not RECAPTCHA_USE_SSL:
+        raise ImproperlyConfigured('Recaptcha is not configured over SSL')
+
+    COOLOFF_TIME = int(get_cooloff_time())
+    LOGIN_FAILURE_LIMIT = int(get_login_failure_limit())
+    LOCKOUT_URL = str(get_lockout_url())
+    if not LOCKOUT_URL or LOCKOUT_URL != 'locked/':
```
Comment (waseem-omar) [inline accountsplus/settings.py]: @nasief This way we are forcing users to set LOCKOUT_URL to `locked/` which doesn't make sense at all. We should have a default value of `locked/` and at the same time we let users to change this value, and add the new value to their `urls.py`

File: accountsplus/views.py [hunk 5]
```diff
@@ -17,6 +18,10 @@
 import django.core.urlresolvers
 
 import signals
+import forms
+import settings
+
+from axes import utils
```
Comment (waseem-omar) [inline accountsplus/views.py]: @nasief This import should go above the local imports.

File: accountsplus/views.py [hunk 6]
```diff
@@ -1,6 +1,7 @@
 from __future__ import unicode_literals
 import logging
 from django.utils.translation import ugettext as _
```
Comment (waseem-omar) [inline accountsplus/views.py]: @nasief django imports should be in its own group

File: requirements.txt [hunk 7]
```diff
@@ -7,3 +7,5 @@ django-timezone-field>=2.0rc1
 django-localflavor
 bcrypt==3.1.0
 mock
+git+https://github.com/foundertherapy/django-axes.git@remove_dependency_on_ip
```
Comment (waseem-omar) [inline requirements.txt:10]: @nasief Shouldn't we add those libraries as dependencies in `setup.py`?

File: README.rst (hunk 1 above)
Comment (nasief) [reply README.rst:39]: Done!

File: README.rst (hunk 2 above)
Comment (nasief) [reply README.rst]: Done!

File: accountsplus/settings.py (hunk 3 above)
Comment (nasief) [reply accountsplus/settings.py]: Done!

File: accountsplus/settings.py (hunk 4 above)
Comment (nasief) [reply accountsplus/settings.py]: Done!

File: accountsplus/views.py (hunk 5 above)
Comment (nasief) [reply accountsplus/views.py]: Done!

File: accountsplus/views.py (hunk 6 above)
Comment (nasief) [reply accountsplus/views.py]: Done!

File: accountsplus/forms.py [hunk 8]
```diff
@@ -0,0 +1,7 @@
+import django.forms
```
Comment (GhalebKhaled) [inline accountsplus/forms.py:3]: @nasief its a good practice to add
`from __future__ import unicode_literals`
At the top of every python files

File: accountsplus/settings.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+# Default values
+LOCKOUT_TEMPLATE = 'accounts/locked_out.html'
+
+def get_setting(setting_str, is_required, default_value=None):
+    try:
+        return getattr(settings, setting_str)
+    except AttributeError:
+        if is_required:
+            raise ImproperlyConfigured('Setting {0} is not configured in your settings module'.format(setting_str))
```
Comment (GhalebKhaled) [inline accountsplus/settings.py]: You can use `{}` instead `{0}` as the `{0}` not needed

File: accountsplus/templates/accounts/locked_out.html [hunk 10]
```diff
@@ -0,0 +1,31 @@
+{% load bootstrap3 %}
+
+{% block content %}
+
+    <div class="col-lg-6 col-lg-offset-3 col-md-6 col-md-offset-3 locked-out-container">
+        <div class="login-message">
+            Your login is blocked due to multiple failing login attempts. Solve the following captcha to get unblocked!
```
Comment (GhalebKhaled) [inline accountsplus/templates/accounts/locked_out.html]: @nasief I suggest to put this string in `blocktrans` so if someone uses this library, he/she can translate these values

File: accountsplus/urls.py [hunk 11]
```diff
@@ -6,8 +6,8 @@
 
 import views
 
-
 urlpatterns = [
+
```
Comment (GhalebKhaled) [inline accountsplus/urls.py]: why we need this extra line? also why you remove the previous one?

File: setup.py [hunk 12]
```diff
@@ -34,6 +34,7 @@
         'django-timezone-field>=2.0rc1',
         'django-localflavor>=1.1',
         'swapper>=0.3.0',
+        'django-recaptcha',
```
Comment (GhalebKhaled) [inline setup.py:37]: need also the `axes` as Waseem mentioned.
I also suggest to add version lock for the minimum`django-recaptcha` required

File: accountsplus/forms.py (hunk 8 above)
Comment (nasief) [reply accountsplus/forms.py:3]: Thanks :D

File: accountsplus/settings.py (hunk 9 above)
Comment (nasief) [reply accountsplus/settings.py]: OK! but still we would need this kind of formatting when we have multiple values to be formatted within the list, right?

File: accountsplus/templates/accounts/locked_out.html (hunk 10 above)
Comment (nasief) [reply accountsplus/templates/accounts/locked_out.html]: you mean `trans` ?

File: setup.py (hunk 12 above)
Comment (nasief) [reply setup.py:37]: Yes, you are right. However, `axes` library we use is not the one on PyPi, we are currently using a forked customized version of this library to resolve an issue with IP tracking. We need to talk to dana about how to resolve such an issue. I am thinking about creating our own version of this library and publish it to PyPi as it is licensed under MIT

File: accountsplus/settings.py (hunk 9 above)
Comment (GhalebKhaled) [reply accountsplus/settings.py]: no, it will be replaced in order

File: accountsplus/urls.py (hunk 11 above)
Comment (nasief) [reply accountsplus/urls.py]: ahaaa!

File: accountsplus/settings.py (hunk 9 above)
Comment (nasief) [reply accountsplus/settings.py]: @GhalebKhaled that's nice!

File: accountsplus/templates/accounts/locked_out.html (hunk 10 above)
Comment (GhalebKhaled) [reply accountsplus/templates/accounts/locked_out.html]: well, both do the same job here so you can you what it fits you more, check this
http://stackoverflow.com/questions/17621770/django-i18n-blocktrans-vs-trans

File: accountsplus/settings.py [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+def get_login_failure_limit():
+    return get_setting('AXES_LOGIN_FAILURE_LIMIT', False, 3)
+
+
+def get_lockout_url():
+    return get_setting('AXES_LOCKOUT_URL', False, 'locked/')
+
+
+def get_lockout_template():
+    return get_setting('TEMPLATE_PATH', False, LOCKOUT_TEMPLATE)
```
Comment (GhalebKhaled) [inline accountsplus/settings.py]: @nasief If `TEMPLATE_PATH` is used only to set lockout template path, then it should be called something like `LOCKOUT_TEMPLATE_PATH`

Comment (nasief) [reply accountsplus/settings.py]: makes sense!

File: accountsplus/settings.py [hunk 14]
```diff
@@ -41,7 +41,7 @@ def get_lockout_url():
 
 
 def get_lockout_template():
-    return get_setting('TEMPLATE_PATH', False, LOCKOUT_TEMPLATE)
+    return get_setting('LOCKOUT_TEMPLATE_PATH', False, LOCKOUT_TEMPLATE)
```
Comment (waseem-omar) [inline accountsplus/settings.py:44]: @nasief I think this will require an update to the instructions in README.

**Pull request merged**